src/stream/turn_server.rs

```diff
@@ -0,0 +1,216 @@
+use std::collections::HashMap;
+use std::net::{IpAddr, Ipv4Addr, SocketAddr};
+use std::sync::{Arc, Mutex};
+use std::thread;
+
+use tokio::net::UdpSocket;
+use tokio::runtime::Runtime;
+
+use util::vnet::net::Net;
+
+use turn::auth::*;
+use turn::relay::relay_static::*;
+use turn::server::{config::*, *};
+use turn::Error;
+
+use log::*;
+
+#[allow(dead_code)]
+pub struct TurnServer {
+    server: Option<turn::server::Server>,
+    public_ip: String,
+    port: u16,
+    realm: String,
+    username: String,
+    password: String,
+    run: bool,
+    running_peers: u32,
+    main_loop_thread: Option<std::thread::JoinHandle<()>>,
+    main_loop_thread_rx_channel: std::sync::mpsc::Receiver<String>,
+}
+
+struct MyAuthHandler {
+    cred_map: HashMap<String, Vec<u8>>,
+}
+
+impl MyAuthHandler {
+    fn new(cred_map: HashMap<String, Vec<u8>>) -> Self {
+        MyAuthHandler { cred_map }
+    }
+}
+
+impl AuthHandler for MyAuthHandler {
+    fn auth_handle(
+        &self,
+        username: &str,
+        _realm: &str,
+        _src_addr: SocketAddr,
+    ) -> Result<Vec<u8>, Error> {
+        if let Some(pw) = self.cred_map.get(username) {
+            debug!("username={}, password={:?}", username, pw);
+            Ok(pw.to_vec())
+        } else {
+            Err(Error::ErrFakeErr)
+        }
+    }
+}
+
+lazy_static! {
+    pub static ref TURN_SERVER: Arc<Mutex<TurnServer>> =
+        Arc::new(Mutex::new(TurnServer::default()));
+}
+
```

Comment:
We should have a public function to access the url of the turn server, with that we can avoid the hardcoded url around the source code for it.